speechbrain/core.py

```diff
@@ -1024,60 +1093,92 @@ def fit_batch(self, batch):
-
-            should_step = self.valid_step % self.grad_accumulation_factor == 0
-            if valid_loss:
-                with self.no_sync(not should_step):
-                    (loss / self.grad_accumulation_factor).backward()
-                if should_step:
-                    self.optimizer.step()
-                    self.zero_grad()
-                    self.optimizer_step += 1
+                outputs = self.compute_forward(batch, sb.Stage.TRAIN)
+
+            # The objectives are removed from Autocast to avoid
+            # potential numerical instabilities with CTC loss, etc.
+            loss = self.compute_objectives(outputs, batch, sb.Stage.TRAIN)
+
+            scaled_loss = self.scaler.scale(
+                loss / self.grad_accumulation_factor
+            )
+            scaled_loss.backward()
+
+        if should_step:
+            self.optimizers_step()
 
         self.on_fit_batch_end(batch, outputs, loss, should_step)
         return loss.detach().cpu()
 
+    def check_gradients(self):
+        """ Checks if the gradients are finite. If not, it will emit a warning and set them to zero."""
+        for param in self.modules.parameters():
+            if param.requires_grad and param.grad is not None:
+                if not torch.isfinite(param.grad).all():
+                    param.grad = None
+                    logger.warning(
+                        f"Gradients {param.name} contain NaN or Inf. Setting to None."
+                    )
+
+    def freeze_optimizers(self, optimizers):
+        """By default, this method returns the passed optimizers.
+        Override this method if you want to freeze some optimizers
+        during training. To do so, return a of active optimizers.
+        """
+        return optimizers
+
+    def on_optimizers_step_end(self):
+        """Gets called at the end of ``optimizers_step()``.
+        Override this method to do something at the end of ``optimizers_step()``.
+        """
+        pass
+
+    def optimizers_step(self):
+        """Performs a step of gradient descent on the optimizers. This method
+        is called every ``grad_accumulation_factor`` steps."""
+        # 1. get the valid optimizers, i.e., the ones that are not frozen during this step
+        valid_optimizers = self.freeze_optimizers(self.optimizers_dict)
+
+        # 2. unscale the gradients of the valid optimizers
+        for opt in valid_optimizers.values():
+            self.scaler.unscale_(opt)
+
+        # 3. clip gradients
+        for opt in valid_optimizers.values():
+            torch.nn.utils.clip_grad_norm_(
+                opt.param_groups[0]["params"], self.max_grad_norm
+            )
+
+        # no need to activate this flag if you are in fp16
+        # since GradScaler is automatically handling the nonfinite gradients
+        if not self.scaler.is_enabled() and self.skip_nonfinite_grads:
+            self.check_gradients()
+
+        # 4. step the valid optimizers
```

Comment:
Can we add a note here that if the scaler is disabled, it simply calls the underlying optimizer's step function? I was confused at first about what was going on in the normal (non-scaled) case.